### PR TITLE
Ic3sa justify coi

### PR DESCRIPTION
**Pull request by makaimann**
Implements the `justify_coi` algorithm for predecessor generalization.

### CODE REVIEW

**Comment by makaimann** [comment]
I don't think this needs a thorough review since it's just getting merged into another PR. But I thought I'd show you this diff separately and post the results. Even in debug mode it's solving more than the previous version of `ic3sa` so it seems like a clear win.

Screenshot from 2021-01-13 11-14-17

**Comment by makaimann** [comment]
I think I've fixed the issue. I just forgot to include regular `constraints` in the `justify_coi` procedure. There are still some issues in `IC3SA` but I think they're unrelated to the changes here. It's only showing up now because it's faster and gets further I believe.

**Comment by ahmed-irfan** [comment]
> I don't think this needs a thorough review since it's just getting merged into another PR. But I thought I'd show you this diff separately and post the results. Even in debug mode it's solving more than the previous version of `ic3sa` so it seems like a clear win. However there are still a bunch of errors so I'll mark it `do not merge` for now.
> 
> Screenshot from 2021-01-13 11-14-17

there are some errors. Do you know more about them?

**Comment by makaimann** [comment]
> > I don't think this needs a thorough review since it's just getting merged into another PR. But I thought I'd show you this diff separately and post the results. Even in debug mode it's solving more than the previous version of `ic3sa` so it seems like a clear win. However there are still a bunch of errors so I'll mark it `do not merge` for now.
> > Screenshot from 2021-01-13 11-14-17
> 
> there are some errors. Do you know more about them?

Yeah it was the `constraints` issue mentioned above. I can run it again, but I'm expecting there to still be errors (I ran into these locally). However, those errors seem unrelated to this `justify_coi` change. Rather, they're related to refinement which I wasn't sure was correct anyway.

File: engines/ic3sa.cpp [hunk 1]
```diff
@@ -37,6 +37,21 @@ using namespace std;
 
 namespace pono {
 
+unordered_set<PrimOp> logical_ops({ And,
+                                    Or,
+                                    // include bit-vector versions for boolector
```
Comment (ahmed-irfan) [inline engines/ic3sa.cpp:42]: implies?

File: engines/ic3sa.cpp [hunk 2]
```diff
@@ -676,6 +722,116 @@ bool IC3SA::add_to_term_abstraction(const Term & term)
...
+    t = to_visit_.back();
+    to_visit_.pop_back();
+
+    if (visited_.find(t) != visited_.end()) {
+      continue;
+    }
+    visited_.insert(t);
+
+    if (t->get_op() == Ite)
+    {
```
Comment (ahmed-irfan) [inline engines/ic3sa.cpp:749]: format

File: engines/ic3sa.cpp [hunk 3]
```diff
@@ -676,6 +722,116 @@ bool IC3SA::add_to_term_abstraction(const Term & term)
...
+      children.clear();
+      children.insert(children.end(), t->begin(), t->end());
+      assert(children.size() == 3);
+      // always visit the condition
+      to_visit_.push_back(children[0]);
+      if (solver_->get_value(children[0]) == solver_true_)
+      {
```
Comment (ahmed-irfan) [inline engines/ic3sa.cpp:756]: format

File: engines/ic3sa.cpp [hunk 4]
```diff
@@ -676,6 +722,116 @@ bool IC3SA::add_to_term_abstraction(const Term & term)
...
+        // the if branch is active
+        to_visit_.push_back(children[1]);
+      }
+      else
+      {
```
Comment (ahmed-irfan) [inline engines/ic3sa.cpp:761]: format

File: engines/ic3sa.cpp [hunk 5]
```diff
@@ -676,6 +722,116 @@ bool IC3SA::add_to_term_abstraction(const Term & term)
...
+      }
+
+    } else {
+      for (const auto & tt : t) {
+        to_visit_.push_back(tt);
+      }
+    }
+  }
+
+  const UnorderedTermSet & inputvars = ts_.inputvars();
```
Comment (ahmed-irfan) [inline engines/ic3sa.cpp:804]: is it used?

File: engines/ic3sa.cpp (hunk 1 above)
Comment (makaimann) [reply engines/ic3sa.cpp:42]: Good point! Addressed in 6740d8fc96e23401e0dde246f9153f50f924be96

File: engines/ic3sa.cpp (hunk 5 above)
Comment (makaimann) [reply engines/ic3sa.cpp:804]: Nope, good catch! 9c956be8b384ec299f4b614613ffb7af59cdf1fd

File: engines/ic3sa.cpp (hunk 2 above)
Comment (makaimann) [reply engines/ic3sa.cpp]: done with 8d265662a1aec3af0a046bf789a64bb66771cf09

**Review by ahmed-irfan — APPROVED** [review]

**Comment by makaimann** [comment]
Thanks for the review!

**Pull request merged**